crates/codegen/parser/src/code_generator.rs

```diff
@@ -102,136 +102,149 @@ impl CodeGenerator {
-            .flatten()
-            .collect::<BTreeSet<_>>()
+    pub fn version_flag_initializers(&self) -> TokenStream {
+        let bindings = self.scanner_and_parser_version_names();
+        let initializers = bindings
             .iter()
             .skip(1) // Don't need the first version
             .map(|version| {
                 let version = version.to_string();
                 let version_name = format_ident!(
-                    "version_is_equal_to_or_greater_than_{}",
-                    version.replace(".", "_")
+                    "version_is_equal_to_or_greater_than_{version}",
+                    version = version.replace(".", "_")
                 );
-                quote! { #version_name: Version::parse(#version).unwrap() <= version }.to_string()
-            })
-            .collect()
+                quote! { #version_name: Version::parse(#version).unwrap() <= version }
+            });
+        quote! { #(#initializers),* }
     }
 
-    pub fn scanner_functions(&self) -> Vec<String> {
-        self.scanners
+    pub fn scanner_functions(&self) -> String {
+        let functions = self
+            .scanners
             .iter()
             .map(|(name, scanner)| {
-                let function_name = format_ident!("scan_{}", name.to_snake_case());
+                let function_name = format_ident!("scan_{name}", name = name.to_snake_case());
                 let body = scanner.to_function_body();
-                format!(
-                    "{}\n{}",
-                    scanner.comment,
-                    quote! {
-                        #[allow(unused_assignments, unused_parens)]
-                        pub(crate) fn #function_name(&self, stream: &mut Stream) -> bool {
-                            #body
-                        }
+                let comment = &scanner.comment;
+                let function = quote! {
+                    #[allow(unused_assignments, unused_parens)]
+                    pub(crate) fn #function_name(&self, stream: &mut Stream) -> bool {
+                        #body
                     }
-                )
+                };
+                format!("{comment}\n{function}")
             })
-            .collect()
+            .collect::<Vec<_>>();
+        functions.join("\n\n")
     }
 
-    pub fn parser_functions(&self) -> Vec<String> {
-        self.parsers
+    pub fn scanner_invocations(&self) -> TokenStream {
+        let invocations = self
+            .scanners
+            .keys()
+            .map(|name| {
+                let production_kind = format_ident!("{name}");
+                let function_name = format_ident!("scan_{name}", name = name.to_snake_case());
+                quote!{ ProductionKind::#production_kind => call_scanner(self, input, Language::#function_name, TokenKind::#production_kind, #name) }
+            });
+        quote! { #(#invocations),* }
+    }
+
+    pub fn parser_functions(&self) -> String {
```

Comment:
Is there a reason `parser_functions` is public?

Reply:
#234

Reply:
Probably none should be public